ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/damle/SandboxDamle.scala

```diff
@@ -36,7 +41,23 @@ object SandboxDamle {
     def resolveStep(result: Result[A]): Future[Either[DamlLfError, A]] = {
       result match {
         case ResultNeedPackage(packageId, resume) =>
-          getPackage(packageId).flatMap(mbPkg => resolveStep(resume(mbPkg)))
+          var gettingPackage = false
+          packagePromises
+            .computeIfAbsent(packageId, { _ =>
```

Comment:
Do you think it would be possible for a call to this method that fails to fetch a package due to a transient error to end up removing a package that has been concurrently fetched on a different thread? Do you think there would be a way to not store failures? We would not longer have to add the removal step and probably synchronizing would be easier.

Reply:
With this change there shouldn't be concurrent package fetching/loading anymore.

Reply:
Except, this only covers the fetching of the package, but not duplicate compilation. Compilation is done of the `needPackage` resume step. Good catch @stefanobaghino-da!

Reply:
Duplicate compilation is already taken care of in ConcurrentCompiledPackages (it's wrapped in `this.synchronized`).